plugins/core/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/telemetry/otel/OtelBaseTest.kt

```diff
@@ -0,0 +1,325 @@
+            val parent = spans.first()
+            val child = spans.last()
+
+            assertThat(parent.parentSpanContext.traceId).isEqualTo(TraceId.getInvalid())
+            assertThat(child.parentSpanContext.traceId).isEqualTo(parent.spanContext.traceId)
+            assertThat(parent.getAttribute(PLUGIN_ATTRIBUTE_KEY)).isEqualTo("Amazon Q For VS Code")
+            assertThat(child.getAttribute(PLUGIN_ATTRIBUTE_KEY)).isEqualTo(parent.getAttribute(PLUGIN_ATTRIBUTE_KEY))
+        }
+    }
+
+    @Test
+    fun `context propagates from parent to child when child#end is after parent#end`() {
+        spanBuilder("tracer", "parentSpan").setParent(Context.current().with(AWS_PRODUCT_CONTEXT_KEY, AWSProduct.AMAZON_Q_FOR_VS_CODE)).use {
+            pluginAwareExecuteOnPooledThread {
+                spanBuilder("anotherTracer", "childSpan").use {
+                    Thread.sleep(100)
+                }
+            }
+        }
+        spinUntil(java.time.Duration.ofSeconds(10)) {
+            otelExtension.completedSpans.size == 2
+        }
+
+        assertThat(otelExtension.completedSpans).hasSize(2).satisfiesKt { spans ->
+            val parent = spans.last()
+            val child = spans.first()
+
+            assertThat(parent.hasEnded())
+            assertThat(child.hasEnded())
+
+            // child started after parent
+            assertThat(child.toSpanData().startEpochNanos).isGreaterThanOrEqualTo(parent.toSpanData().startEpochNanos)
+            // and called end after parent
+            assertThat(child.toSpanData().endEpochNanos).isGreaterThanOrEqualTo(parent.toSpanData().endEpochNanos)
+
+            assertThat(parent.parentSpanContext.traceId).isEqualTo(TraceId.getInvalid())
+            assertThat(child.parentSpanContext.traceId).isEqualTo(parent.spanContext.traceId)
+            assertThat(parent.getAttribute(PLUGIN_ATTRIBUTE_KEY)).isEqualTo("Amazon Q For VS Code")
+            assertThat(child.getAttribute(PLUGIN_ATTRIBUTE_KEY)).isEqualTo(parent.getAttribute(PLUGIN_ATTRIBUTE_KEY))
+        }
+    }
+
+    private fun spanBuilder(tracer: String, spanName: String) = DefaultSpanBuilder(otelExtension.sdk.sdk.getTracer(tracer).spanBuilder(spanName))
+}
+
+class OtelExtension : AfterEachCallback, AfterAllCallback {
+    private val openSpans = mutableSetOf<ReadableSpan>()
+    private val _completedSpans = mutableListOf<ReadableSpan>()
+    val completedSpans
+        get() = _completedSpans.reversed().toList()
+
+    val sdk = OTelService(
+        listOf(
+            // should probably be a service loader
+            object : SpanProcessor {
+                override fun isStartRequired() = true
+                override fun isEndRequired() = true
+
+                override fun onStart(parentContext: Context, span: ReadWriteSpan) {
+                    openSpans.add(span)
+                }
+
+                override fun onEnd(span: ReadableSpan) {
+                    println(span)
```

Comment:
do we need this?

Reply:
helpful for debugging without a debugger